server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java

```diff
@@ -0,0 +1,388 @@
+    }
+  }
+
+  private boolean readAndCalcFromChunk(long curStartTime, long curEndTime)
+      throws IOException, QueryProcessException {
+    while (reader.hasNextChunk()) {
+      // try to calc from chunkMetaData
+      Statistics chunkStatistics = reader.currentChunkStatistics();
+      if (chunkStatistics != null && !isEmpty(chunkStatistics)) {
+        if (chunkStatistics.getStartTime() >= curEndTime) {
+          if (ascending) {
+            return true;
+          } else {
+            reader.skipCurrentChunk();
+            continue;
+          }
+        }
+        if (reader.canUseCurrentChunkStatistics()
+            && timeRange.contains(chunkStatistics.getStartTime(), chunkStatistics.getEndTime())) {
+          // calc from chunkMetaData
+          while (reader.hasNextSubSeries()) {
+            Statistics currentChunkStatistics = reader.currentChunkStatistics();
+            calcFromStatistics(currentChunkStatistics, results.get(reader.getCurIndex()));
+            reader.nextSeries();
+          }
+          reader.skipCurrentChunk();
+          continue;
+        }
+      }
+      // read page
+      if (readAndCalcFromPage(curStartTime, curEndTime)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean readAndCalcFromPage(long curStartTime, long curEndTime)
+      throws IOException, QueryProcessException {
+    while (reader.hasNextPage()) {
+      // try to calc from pageHeader
+      Statistics pageStatistics = reader.currentPageStatistics();
+      // must be non overlapped page
+      if (pageStatistics != null && !isEmpty(pageStatistics)) {
+        // current page max than time range
+        if (pageStatistics.getStartTime() >= curEndTime) {
+          if (ascending) {
+            return true;
+          } else {
+            reader.skipCurrentPage();
+            continue;
+          }
+        }
+        if (reader.canUseCurrentPageStatistics()
+            && timeRange.contains(pageStatistics.getStartTime(), pageStatistics.getEndTime())) {
+          // calc from pageHeader
+          while (reader.hasNextSubSeries()) {
+            Statistics currentPageStatistics = reader.currentPageStatistics();
+            calcFromStatistics(currentPageStatistics, results.get(reader.getCurIndex()));
+            reader.nextSeries();
+          }
+          reader.skipCurrentPage();
+          if (isEndCalc()) {
+            return true;
+          }
+          continue;
+        }
+      }
+
+      // calc from page data
+      BatchData batchData = reader.nextPage();
+      if (batchData == null || !batchData.hasCurrent()) {
+        continue;
+      }
+
+      // stop calc and cached current batchData
+      if (ascending && batchData.currentTime() >= curEndTime) {
+        preCachedData = batchData;
+        // reset the last position to current Index
+        lastReadCurArrayIndex = batchData.getReadCurArrayIndex();
+        lastReadCurListIndex = batchData.getReadCurListIndex();
```

Comment:
put this before the if block and delete the following same two lines, just like `LocalGroupByExecutor`

Reply:
fixed